e2e/ctl_v3_alarm_test.go

```diff
@@ -99,6 +108,101 @@ func alarmTest(cx ctlCtx) {
 	}
 }
 
+type fakeConsistentIndex struct{ rev uint64 }
+
+func (f *fakeConsistentIndex) ConsistentIndex() uint64 { return f.rev }
+
+func alarmCorruptTest(cx ctlCtx) {
+	for i := 0; i < 10; i++ {
+		if err := ctlV3Put(cx, "k", "v", ""); err != nil {
+			if cx.dialTimeout > 0 && !isGRPCTimedout(err) {
+				cx.t.Fatalf("putTest ctlV3Put error (%v)", err)
+			}
+		}
+	}
+	cx.epc.procs[0].Stop()
+
+	// Corrupt member 0 by modifying backend offline.
+	fp := filepath.Join(cx.epc.procs[0].Config().dataDirPath, "member", "snap", "db")
+	be := backend.NewDefaultBackend(fp)
+	s := mvcc.NewStore(be, nil, &fakeConsistentIndex{13})
+	s.Put([]byte("abc"), []byte("def"), 0)
+	s.Put([]byte("xyz"), []byte("123"), 0)
+	s.Compact(5)
+	s.Commit()
+	s.Close()
+	be.Close()
+
+	eps := cx.epc.EndpointsV3()
+
+	// Wait for cluster so Puts succeed in case member 0 was the leader.
+	cli1, err := clientv3.New(clientv3.Config{Endpoints: []string{eps[1]}, DialTimeout: 3 * time.Second})
+	if err != nil {
+		cx.t.Fatal(err)
+	}
+	defer cli1.Close()
+
+	if _, err = cli1.Get(context.TODO(), "k"); err != nil {
+		cx.t.Fatal(err)
+	}
+	cli1.Put(context.TODO(), "xyz", "321")
+	cli1.Put(context.TODO(), "abc", "fed")
+
+	// Restart with corruption checking
+	cx.epc.procs[1].Stop()
+	cx.epc.procs[2].Stop()
+	cx.epc.Restart()
+
+	// wait until corruption detected
+	cli0, err := clientv3.New(clientv3.Config{Endpoints: []string{eps[0]}, DialTimeout: 3 * time.Second})
+	if err != nil {
+		cx.t.Fatal(err)
+	}
+	defer cli0.Close()
+
+	sresp, err := cli1.Status(context.TODO(), eps[0])
+	if err != nil {
+		cx.t.Fatal(err)
+	}
+	id0 := sresp.Header.GetMemberId()
+
+	corrupted := false
+	for i := 0; i < 5; i++ {
+		presp, perr := cli0.Put(context.TODO(), "abc", "aaa")
+		if perr != nil {
+			if perr.Error() != rpctypes.ErrCorrupt.Error() {
+				cx.t.Fatalf("expected %v, got %+v (%v)", rpctypes.ErrCorrupt, presp, perr)
+			}
+			corrupted = true
+			break
+		}
+		time.Sleep(time.Second)
+	}
+	if !corrupted {
+		cx.t.Fatalf("expected error %v after %s", rpctypes.ErrCorrupt, 5*time.Second)
+	}
+
+	// corrupt alarm should now be on
```

Comment:
The cluster that's currently running can be expected to be OK since it's periodically checking its hash with its other members. The newly corrupted member should fatal out before joining raft / doing any damage / causing need for an alarm.

Reply:
> The newly corrupted member should fatal out before joining raft

> Instead, if a member can contact the client addresses of its peers, it should first fetch hashes from the other members at a known revision and compare before serving any client requests. (from original issue)

Agree.

So should the checking be before `s.publish` request, in `s.start`, using Raft entries?

https://github.com/coreos/etcd/blob/master/etcdserver/server.go

```go
func (s *EtcdServer) Start() {
	s.start()
	s.goAttach(func() { s.publish(s.Cfg.ReqTimeout()) })
	s.goAttach(s.purgeFile)
	s.goAttach(func() { monitorFileDescriptor(s.stopping) })
	s.goAttach(s.monitorVersions)
	s.goAttach(s.linearizableReadLoop)
	s.goAttach(s.monitorKVHash)
}
```

I am trying to figure out the best way to fetch hashes from other members without starting gRPC server, or starting Raft. Or on restart, we still start Raft and gRPC server but do not accept new requests other than hash kv RPC?
> The newly corrupted member should fatal out before joining raft

> Instead, if a member can contact the client addresses of its peers, it should first fetch hashes from the other members at a known revision and compare before serving any client requests. (from original issue)

Agree.

So should the checking be before `s.publish` request, in `s.start`, using Raft entries?

https://github.com/coreos/etcd/blob/master/etcdserver/server.go

```go
func (s *EtcdServer) Start() {
	s.start()
	s.goAttach(func() { s.publish(s.Cfg.ReqTimeout()) })
	s.goAttach(s.purgeFile)
	s.goAttach(func() { monitorFileDescriptor(s.stopping) })
	s.goAttach(s.monitorVersions)
	s.goAttach(s.linearizableReadLoop)
	s.goAttach(s.monitorKVHash)
}
```

I am trying to figure out the best way to fetch hashes from other members without starting gRPC server, or starting Raft. Or on restart, we still start Raft and gRPC server but do not accept new requests other than hash kv RPC?

Reply:
Probably just before `s.start()` entirely; load the cluster membership info and hash from the backend, issue the hashkv RPCs to the member addresses and compare with the backend, then call start() if there's no mismatch.